mantidproject/mantid
# Fix Unreliable Access Violation when using QtPropertyBrowser

**Description of work.**
Credit to @martyngigg for helping fix this problem.

This PR fixes an unreliable access violation when opening the `FunctionTreeView`. A similar issue is also present in `FitOptionsBrowser`. It is suspected that the issue might still be present in other areas where `unsetFactoryForManager` is not used in the destructors, there is a separate issue for this #30419.

Its cause was a dangling pointer to a manager object being left behind when destructing a FunctionTreeView. This dangling point was still existing in a global static variable (m_managerToFactoryToViews or m_viewToManagerToFactory) in qtpropertybrowser.cpp. When creating a new instance of FunctionTreeView, the memory location would sometimes be reused,
causing problems.

The solution used to fix this was to do:
```
m_browser->unsetFactoryForManager(m_manager)
```
in the destructor of FunctionTreeView and FitOptionsBrowser.
 
*Release notes are not needed as this is a very unlikely error to happen - it probably hasn't been noticed.*

**To test:**
It has been confirmed that the new `FunctionTreeViewTest` and `FitOptionsBrowserTest` will have previously failed in an unreliable way - these should no longer fail.

Run the `FunctionTreeViewTest`, `FitOptionsBrowserTest` and `FitScriptGeneratorViewTest` many times and make sure it does not fail.
```
ctest --repeat-until-fail 25 -C RelWithDebInfo -R MantidQtWidgetsCommonTestQt5_FitScriptGeneratorViewTest -V
```

A quick test of the indirect data analysis interface is also recommended.

Fixes #30362

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
